CRM/Contribute/BAO/Premium.php

```diff
@@ -104,32 +104,25 @@ public static function del($premiumID) {
   public static function buildPremiumBlock(&$form, $pageID, $formItems = FALSE, $selectedProductID = NULL, $selectedOption = NULL) {
     $form->add('hidden', "selectProduct", $selectedProductID, array('id' => 'selectProduct'));
 
-    $dao = new CRM_Contribute_DAO_Premium();
-    $dao->entity_table = 'civicrm_contribution_page';
-    $dao->entity_id = $pageID;
-    $dao->premiums_active = 1;
-    CRM_Financial_BAO_FinancialType::getAvailableFinancialTypes($financialTypes, CRM_Core_Action::ADD);
-    $addWhere = "financial_type_id IN (0)";
-    if (!empty($financialTypes)) {
-      $addWhere = "financial_type_id IN (" . implode(',', array_keys($financialTypes)) . ")";
-    }
```

Comment:
This required by Financial Type ACL when access control by financial type is enabled at Administer > CiviContribute Component Settings. The fix would be to check if financial type is NULL. i.e

```
--- a/CRM/Contribute/BAO/Premium.php
+++ b/CRM/Contribute/BAO/Premium.php
@@ -113,7 +113,7 @@ class CRM_Contribute_BAO_Premium extends CRM_Contribute_DAO_Premium {
     if (!empty($financialTypes)) {
       $addWhere = "financial_type_id IN (" . implode(',', array_keys($financialTypes)) . ")";
     }
-
+    $addWhere = "{$addWhere} OR financial_type_id IS NULL";
     if ($dao->find(TRUE)) {
       $premiumID = $dao->id;
       $premiumBlock = array();
```

@Edzelopez @monishdeb your thoughts on this?

Reply:
@pradpnayak Makes sense, I've amended this PR to make the suggested change.

Reply:
@pradpnayak I did move the addWhere code into the if{} below as there's no point retreiving the financial types if we're not going to use them.

Reply:
I'd really prefer we fixed up this function in it's own PR - it's a core BAO fn & it feels like a hidden change

Reply:
@eileenmcnaughton This change is the reason for the PR?  The extra bit is moving the redirect from the BAO to the form layer.  Actual change looks bigger than it is because I've moved the call to getAvailableFinancialTypes() inside the "if" as it's not used otherwise.

Reply:
OK-  I can see now that the other code is supporting code to this now I think